### PR TITLE
refactor(tokenless): migrate dsh lifecycle

### PR DESCRIPTION
**Pull request by boluochoufeng**
## Why

The DSH bundle still owned a parallel JSON compression and environment-diagnostic policy, so its behavior could drift from the PostTool Core used by the other adapters. This change makes DSH the final current adapter to delegate the lifecycle operations supported by its host seam.

## What changed

- Send replaceable single-text results and tool failures through the Core-owned PostTool operation while preserving DSH waterfall decisions and fail-open behavior.
- Keep only DSH status, content-origin, attribution, and projection translation in the bundle; remove local JSON parsing, thresholds, winner selection, error regexes, and compression policy options.
- Declare recovery unavailable, add DSH lifecycle contract coverage, and update the bilingual integration documentation.

## Related issue

no-issue: lifecycle architecture migration

## User / Agent impact

Eligible DSH JSON results now use the same Core cleanup and optional TOON selection as other adapters. File content and lossy candidates remain unchanged because DSH has no Marker-authorized recovery path. The skipTools, shellTools, truncateStringsAt, truncateArraysAt, maxDepth, and noStash plugin options are removed; timeout, buffer, binary path, Agent attribution, and the compression switch remain available.

## Risk and compatibility

- [x] Public CLI, API, configuration, or documented behavior changed
- [ ] Privileged or security-sensitive behavior changed
- [x] Cross-component contract changed
- [x] Migration or rollback guidance is needed

The DSH bundle must be paired with a Tokenless CLI from the same lifecycle-based release. Existing profile overrides should remove the deleted policy options; Core now owns those settings. DSH does not expose BeforeModel, PreTool, or Marker-authorized Retrieve seams, so this PR intentionally migrates PostTool only.

## Validation

- node --check adapters/tokenless/dsh/dist/index.js
- bash tests/test-dsh-adapter.sh
- make test-adapters
- make test-integration
- cargo fmt --all -- --check
- cargo clippy --workspace --all-targets --locked -- -D warnings
- cargo test --workspace -- --test-threads=1
- make test-hooks with the workspace tokenless and packaged RTK on PATH
- make test-raw-package
- make test-npm-package
- make test-toon-cleanup
- Real Core smoke: DSH reduced a 2,221-character JSON result to 598 characters through tokenless compress.

## Documentation and rollback

Updated the English and Chinese Tokenless README, framework integration guide, and user manual. Roll back by reverting commit 10c1cde9 or disabling the DSH adapter; responseCompressionEnabled=false disables successful-result compression while retaining error diagnostics.

### CODE REVIEW

File: src/tokenless/adapters/tokenless/dsh/dist/index.js [hunk 1]
```diff
@@ -14,38 +14,27 @@ const DEFAULT_AGENT_ID = 'dsh'
...
   'notebookread',
+  'notebook_read',
 ])
 
-// These values are the same thresholds used by the shared hook adapter.  A
-// dsh profile can override them in its plugin config; the Tokenless CLI still
-// owns the actual compression algorithm and stats recording.
-const DEFAULT_SHELL_TOOLS = new Set([
-  'Bash',
+const COMMAND_TOOLS = new Set([
```
Comment (ikunkun-sys) [inline src/tokenless/adapters/tokenless/dsh/dist/index.js:36]: **[P1] Preserve failure semantics for the official `pwsh` tool**

`@deepseek-ai/dsh-tool-pwsh@0.1.1-rc.2` is within the declared peer range and is included by the standard DSH presets. It registers the name `pwsh` and returns the same foreground process envelope as `bash`. Because `pwsh` is absent from this set, a result such as `{ exitCode: 1, stderr: "permission denied" }` is sent to Core as `status: "success"`, `content_origin: "api_response"`, and `replace_output: true`. This bypasses failure diagnostics and status accounting, and permits a failed PowerShell command to be replaced as successful output.

File: src/tokenless/adapters/tokenless/dsh/dist/index.js [hunk 2]
```diff
@@ -142,80 +54,74 @@ function valueOr(config, key, fallback) {
...
-    || (/^-?\d+$/.test(stringExitCode) && Number(stringExitCode) !== 0)
-  const timedOut = value.timed_out === true || value.timedOut === true
-  const failed = nonzeroExit
-    || timedOut
+  const numericExit = typeof exitCode === 'number'
+    ? exitCode
+    : (typeof exitCode === 'string' && /^-?\d+$/.test(exitCode.trim())
+        ? Number(exitCode)
+        : 0)
+  return numericExit !== 0
```
Comment (ikunkun-sys) [inline src/tokenless/adapters/tokenless/dsh/dist/index.js:87]: **[P2] Preserve failure semantics for signal-terminated commands**

The official DSH foreground process envelope can report `{ exitCode: null, signal: "SIGTERM", timedOut: false, isError: false }`. Since `commandFailed()` does not inspect `signal`, this result is sent as `status: "success"` with `replace_output: true`. A killed command therefore enters the success/compression path and skips Core's error diagnostics and failure accounting.

**Review by ikunkun-sys — CHANGES_REQUESTED** [review]
Requesting changes for two current-head host-contract regressions: the official `pwsh` command is not classified as a command, and signal-terminated foreground commands are reported as successful. Both were reproduced against the DSH 0.1.1-rc.2 contracts on `10c1cde99e`.

**Review by Forrest-ly — APPROVED** [review]
## 总体评价

DSH 适配器向 PostTool Core 的生命周期迁移干净、完整：协议 v2 请求形态、失败开放（fail-open）语义、waterfall 优先级与诊断上下文结构均与 Core 契约及已迁移的 OpenClaw/Hermes 适配器保持一致；测试覆盖全面，本地实际运行通过。未发现阻塞问题。

## 审查结论

approve

## 详细意见

### 🔴 必须修改（阻塞合并）

无。

核对过的关键契约均正确：
- `protocol_version: 2` / `operation: 'post_tool'` 信封与 `PostToolRequest`（`deny_unknown_fields`）字段完全匹配；`status`（success/error/interrupted）、`content_origin`、`capabilities` 均为 `tokenless-protocol` 合法取值。
- 文件内容无损由 Core 侧保证（`pipeline.rs` 对 `ContentOrigin::FileContent` 直接 passthrough）；`retrieval_available: false` + `require_reversibility: true` 使有损候选被 `RecoverabilityUnavailable` 拒绝，与 PR 描述一致。
- `FILE_TOOLS`/`COMMAND_TOOLS` 与 `tool_categories.json` 的 layer_1/layer_2 集合（小写化后）一致，且测试用 category 文件逐一遍历强制校验，防止漂移。
- 诊断上下文消息结构与旧版 `attributionContext` 完全相同，DSH 宿主契约未变；`additionalContexts` 追加保留了下游上下文。
- 删除的插件选项（skipTools、shellTools、truncateStringsAt、truncateArraysAt、maxDepth、noStash）在代码、安装模板与文档中无残留引用；旧 profile 携带这些键时仅被静默忽略，失败安全。
- 本地验证：`bash tests/test-dsh-adapter.sh` 通过，`node --check dist/index.js` 通过。

### 🟡 建议修改（不阻塞但推荐）

- [`src/tokenless/adapters/tokenless/dsh/dist/index.js:272`] success 分支用 `decision.content ?? result.content`，而同一处理器的错误分支（行 293）用 `decision.content === undefined` 判断。若某个下游 listener 显式把 `content` 置为 `null`（非法但可能发生），success 分支会回退压缩 `result.content`，错误分支则视其为显式投影，两处语义不一致。建议统一为 `=== undefined` 判断（或都用 `??`），保持单一语义。
- [`src/tokenless/adapters/tokenless/dsh/dist/index.js:270-283`] 每次成功的文件读取类工具（read/grep/glob 等 FILE_TOOLS）都会拉起一次 `tokenless compress` 子进程，而 Core 对 `file_content` 是无条件 passthrough（`post_tool/pipeline.rs:74-79`）。旧实现对这类工具完全不产生子进程。这与 OpenClaw 的统一委托形态一致、可能是有意设计，但文件读取是最高频工具类，建议确认是否接受该延迟成本；若想优化，可在宿主映射阶段对已知 `file_content` 来源免 spawn（这并不把策略收回适配器，因为 Core 对该 origin 本来就只有 passthrough 一种结果）。同理，`interrupted` 状态在 Core 侧也是无条件 passthrough 且无诊断（`entry.rs:566-571`），当前同样会触发一次 spawn，可视统计需要决定是否保留。

### 🟢 值得肯定

- 测试非常扎实：覆盖 applied/passthrough、下游 content 投影、value 替换（含结构化命令失败）、原始失败与结构化失败分流、ABORTED/ABORTED_BEFORE_DISPATCH 与 signal 中止、Code Mode 父子执行、压缩开关关闭时诊断仍生效、以及 fail/invalid/timeout/missing/wrong-version/wrong-operation 全部失败开放路径。
- `interrupted` 与 `error` 的区分交由 Core 统一处理（interrupted 透传、不再误加"跳过重试"归因），语义比旧版更正确。
- 中英双语文档（framework-integration、user-manual、README）同步更新，选项表、行为描述与 cordis.patch.yml 示例一致，迁移/回滚指引清晰。
- 归因字段（agent_id/session_id/tool_use_id）构造方式保持稳定，`agentId` 仍可配置且测试验证。

**Review by Forrest-ly — CHANGES_REQUESTED** [review]
## 总体评价

补充审查：在我提交 approve 之后出现的两项 changes-requested 意见（ikunkun-sys）与 Codex bot 的 P2 意见，我逐条对照 DSH 0.1.1-rc.2 官方包与本 PR 之前的实现做了独立验证，三点均属实，且都落在本 PR 新增/改写的代码上。因此撤回我此前的 approve，改为 request changes。

## 审查结论

request changes

## 详细意见

### 🔴 必须修改（阻塞合并）

以下为对已有 review 意见的独立验证与实质性补充（不重复其描述，只补充证据与修法）：

- [`src/tokenless/adapters/tokenless/dsh/dist/index.js:36-46`] **pwsh 缺失（补充 ikunkun-sys 的 P1）**。已核实：`@deepseek-ai/dsh-tool-pwsh@0.1.1-rc.2` 在 npm 上存在且与本适配器声明的 peer 区间（`@deepseek-ai/dsh-tools >=0.1.0-rc.2 <0.2.0`）同版本线，其 `package/lib/index.js` 注册 `name: "pwsh"`（行 234），并通过与内置命令工具相同的 `ctx.shell` 服务返回 `{ kind: "foreground", exitCode, signal, timedOut, aborted, stdout, stderr }` 信封。`pwsh` 不在 `COMMAND_TOOLS` 时，失败的 PowerShell 命令会以 `status: "success"` + `content_origin: "api_response"` + `replace_output: true` 发给 Core，失败诊断被绕过、输出可被当作成功结果替换。修法建议：由于 OpenClaw 等适配器的命令工具集派生自 `adapters/tokenless/common/hooks/tool_categories.json` 的 `layer_2_shell`，建议把 `pwsh` 加入该分类表（DSH 的 `COMMAND_TOOLS` 同步更新），测试对分类表的逐一遍历会自动覆盖；如有其他官方命令类工具（如 `cmd`/`powershell` 形态）也请一并核对。
- [`src/tokenless/adapters/tokenless/dsh/dist/index.js:78-92`] **signal 终止未判失败（补充 ikunkun-sys 的 P2）**。已核实：Node 子进程被信号杀死时 `exitCode` 为 `null`、`signal` 为信号名，DSH 官方信封原样透传这两个字段（见 `dsh-tool-pwsh` 的 `canonicalPwshResult`）。当前 `commandFailed()` 对 `{ exitCode: null, signal: "SIGTERM", timedOut: false }` 返回 false，被杀命令进入成功/压缩路径。修法建议：`commandFailed` 将 `typeof value.signal === 'string' && value.signal.length > 0`（以及可考虑的 `value.aborted === true`）视为失败；注意与 `exec.signal?.aborted` 的早退路径区分（宿主中止仍走 interrupted 语义）。
- [`src/tokenless/adapters/tokenless/dsh/dist/index.js:116-118`] **JSON.stringify 未防护（补充 Codex bot 的 P2）**。已核实为本 PR 引入的回归：迁移前实现的 `classifyStructuredEnvironmentError` 对 `JSON.stringify(error)` 包有 try/catch 并以 `String(error)` 兜底（旧实现约行 197-204），新 `failureText` 去掉了该防护。`value.error` 含循环引用或 BigInt 时会在 `runPostTool` 的 fail-open catch 之外抛出，整个 `tools/post-execute` listener reject，把已完成的工具失败变成插件失败，违反失败开放契约。修法建议：恢复 try/catch + `String(value.error)` 兜底。

### 🟡 建议修改（不阻塞但推荐）

（维持我上一条 review 中的两点，此处不重复：`??` 与 `=== undefined` 的 content 判空不一致（行 272 / 293）；文件类工具与 interrupted 结果触发必然 passthrough 的子进程 spawn。）

### 🟢 值得肯定

- 三个问题均可通过小范围改动修复（分类表 + 两个谓词分支 + 一个 try/catch），且现有测试框架（分类表遍历、fail-open 用例）天然适合为这三点补测试。
- 其余结论不变：协议信封、waterfall 语义、诊断上下文结构、失败开放主路径与双语文档均正确。

File: src/tokenless/adapters/tokenless/dsh/dist/index.js (hunk 1 above)
Comment (boluochoufeng) [reply src/tokenless/adapters/tokenless/dsh/dist/index.js:36]: Fixed in b2de411e. The adapter now classifies the official `pwsh` tool as `command_output`; a regression test verifies a non-zero foreground result is sent to Core as `status: error` with output replacement disabled.

File: src/tokenless/adapters/tokenless/dsh/dist/index.js (hunk 2 above)
Comment (boluochoufeng) [reply src/tokenless/adapters/tokenless/dsh/dist/index.js:87]: Fixed in b2de411e. A non-empty foreground `signal` now marks the command as failed, and the signal is included in the Core diagnostic input so an empty stderr cannot bypass PostTool. The regression covers `exitCode: null`, `SIGTERM`, and empty streams.

File: src/tokenless/adapters/tokenless/dsh/dist/index.js [hunk 3]
```diff
@@ -142,80 +54,74 @@ function valueOr(config, key, fallback) {
...
+      parts.push(stderr)
+    } else if (stderr && typeof stderr.text === 'string') {
+      parts.push(stderr.text)
+    }
+    if (typeof value.error === 'string') {
+      parts.push(value.error)
+    } else if (value.error && typeof value.error.message === 'string') {
+      parts.push(value.error.message)
+    } else if (value.error && typeof value.error === 'object') {
+      parts.push(JSON.stringify(value.error))
```
[the remark this answers is not in the document]
Comment (boluochoufeng) [reply src/tokenless/adapters/tokenless/dsh/dist/index.js]: Fixed in b2de411e. Original tool values are snapshotted as `JsonValue` before this seam, but a downstream waterfall listener can return `decision.value` before DSH revalidates it. The object-to-text boundary now guards `JSON.stringify` locally and falls back to string coercion, so Tokenless does not reject first; a circular downstream value regression verifies that DSH retains ownership of final validation.

File: src/tokenless/adapters/tokenless/dsh/dist/index.js [hunk 4]
```diff
@@ -229,169 +146,181 @@ function attributionContext(text) {
...
+        ...decision,
+        content: [{ type: 'text', text: response.output }],
+      }
     }
+
+    const effectiveResult = decision.content === undefined
+      ? result
+      : { ...result, content: decision.content }
+    const content = structuredFailure
+      ? failureText(undefined, effectiveValue)
```
Comment (ikunkun-sys) [inline src/tokenless/adapters/tokenless/dsh/dist/index.js:315]: **[P2] Keep structured command failures on the Core path**

A valid DSH foreground command can exit non-zero without `stderr`, `error`, or `signal`; for example, a Bash command may print `permission denied` to stdout and then exit 1. `commandFailed()` recognizes the failure, but this call passes only the canonical value to `failureText()`, which ignores stdout and the model-visible `result.content`. With empty stderr it returns an empty string, so line 310 exits before `runPostTool()`. An event-level probe on this head using `exitCode: 1`, stdout `permission denied`, and empty stderr observed no Tokenless invocation. The failure therefore misses Core's `tool_error` routing, environment diagnostic, and accounting, contrary to the updated documentation that structured command failures are sent to Core.

**Review by ikunkun-sys — CHANGES_REQUESTED** [review]
The two findings from my previous review are fixed and revalidated on `b2de411e2b`. Requesting changes for one remaining current-head lifecycle gap: a valid structured command failure with empty stderr can bypass Core entirely.

**Review by Forrest-ly — CHANGES_REQUESTED** [review]
## Re-review of `b2de411e` — the three earlier findings are verified fixed; one remaining lifecycle gap

### Verified fixed

1. **[P1] pwsh classification** — `pwsh` is now in `COMMAND_TOOLS` (`dist/index.js:38`); a non-zero pwsh foreground result is sent to Core as `status: "error"`, `content_origin: "command_output"`, `replace_output: false`. Covered by the new pwsh case in `tests/test-dsh-adapter.sh` (lines 253–267). Keeping the alias DSH-local instead of touching the shared taxonomy is fine, per the author's note.
2. **[P2] signal termination** — `commandFailed()` now treats a non-empty `signal` string as failure (`dist/index.js:88`) and `failureText()` appends `terminated by signal: <SIG>` (`dist/index.js:127–129`), so `{exitCode: null, signal: "SIGTERM"}` no longer enters the success/compression path. Covered by the SIGTERM test (lines 269–281); shell-shaped business values from API tools remain `success` (over-triggering guarded).
3. **[Codex P2] stringify guard** — object `value.error` serialization is wrapped in try/catch with a `String()` fallback (`dist/index.js:119–124`), covering circular references and BigInt; covered by the circular-error test (lines 315–329).

Wire fields match the protocol-v2 `deny_unknown_fields` contract exactly, fail-open is preserved for every transport failure mode, and `bash tests/test-dsh-adapter.sh` plus `make build-dsh-plugin` pass locally on this head. Bilingual docs for the removed options are complete.

### Remaining gap — structured command failure with empty stderr bypasses Core

Independently confirmed; matches @ikunkun-sys' 11:30 review. `dist/index.js:307–310`:

```js
const content = structuredFailure
  ? failureText(undefined, effectiveValue)
  : failureText(effectiveResult)
if (!content) return decision
```

For structured failures only the canonical value is inspected (`stderr` / `error` / `signal`). A command that exits non-zero with its diagnostic text on **stdout** only (e.g. prints `permission denied`, then `exit 1`) yields empty content, and the early return means `runPostTool()` is never called — the failure misses Core's `tool_error` routing, environment diagnosis, and accounting. This contradicts the documentation this PR adds ("Raw DSH failures and structured command failures are sent to Core for environment diagnosis even when compression is disabled", EN/ZH framework-integration.md).

Note the `replacesValue` branch (`dist/index.js:263–266`) shares the same `failureText(undefined, effectiveValue)` + `if (!content) return decision` pattern, so a downstream-replaced canonical value with the same shape also bypasses Core silently.

Suggested direction (preserves existing semantics — stderr/error/signal keep precedence, and stdout must not drive diagnosis when stderr exists):
- When the value-only extraction is empty, fall back to the model-visible rendered output (e.g. `failureText(effectiveResult)`, whose content blocks carry the rendered stdout) so Core still receives text to diagnose;
- Apply the same fallback in the `replacesValue` branch, using `decision.content` when provided so stale display content is not reintroduced;
- Add a regression test: bash with `exitCode: 1`, stdout `permission denied`, empty stderr → assert a PostTool invocation with `status: "error"` whose content includes the stdout text, and that the diagnostic context is attached.

### Minor (non-blocking)

Docs: the plugin silently ignores unknown config keys, so consider one sentence noting that the removed options (`skipTools`, `shellTools`, `truncateStringsAt`, `truncateArraysAt`, `maxDepth`, `noStash`) are no longer read — helpful for users who still have them in existing `cordis.patch.yml` overrides.

File: src/tokenless/adapters/tokenless/dsh/dist/index.js (hunk 4 above)
Comment (boluochoufeng) [reply src/tokenless/adapters/tokenless/dsh/dist/index.js:315]: Fixed in 29e7e911. DSH canonical failures now prefer stderr/error/signal and fall back to stdout when that is the only diagnostic payload. More importantly, every host-classified failure is sent to Core even when all streams are empty, so tool_error routing and accounting cannot be bypassed by missing text. Regression coverage includes both stdout-only and completely empty non-zero foreground results; make test and the adapter suite pass.

**Review by Forrest-ly — CHANGES_REQUESTED** [review]
## 总体评价

对新 head `b2de411` 做了完整复审（含 10c1cde → b2de411 增量核对）。此前三个阻塞问题（pwsh 未分类为命令工具、signal 终止未判失败、JSON.stringify 未防护）均已在当前 head 修复，且各自带针对性回归测试，我逐行核对代码确认修复到位。协议信封、waterfall 语义、失败开放路径与双语文档保持正确。剩余的"空 stderr 结构化命令失败绕过 Core"问题已由 ikunkun-sys 在当前 head 的 review 中指出；我的审查确认其属实，并补充一个具有相同缺陷的第二调用点。

## 审查结论

request changes

## 详细意见

### 🔴 必须修改（阻塞合并）

- [`src/tokenless/adapters/tokenless/dsh/dist/index.js:259-260`] **（补充 ikunkun-sys review）** 空载荷绕过同样存在于下游替换值（replacesValue）分支。若后续 waterfall listener 把命令工具的 canonical `value` 替换为不带 `stderr`/`error`/`signal` 的失败形状（例如 `{ kind: 'foreground', exitCode: 1, stdout: { text: 'permission denied', truncated: false }, stderr: { text: '', truncated: false } }`），`commandFailed()` 判定失败，但 `failureText(undefined, effectiveValue)` 返回空字符串，处理器在 `runPostTool()` 之前返回，该失败同样不会到达 Core 的 tool_error 路由、环境诊断与统计。ikunkun-sys 已指出非替换分支（行 307-310）的同一问题；修复时请让两个调用点获得相同的兜底来源（stdout / 模型可见 content——替换分支的模型可见内容为 `decision.content ?? result.content`），并保持 stderr 优先顺序（现有测试 "command stdout must not drive diagnosis" 钉住了该优先级）。建议同时补一条"替换值 + 空 stderr"的回归。

### 🟡 建议修改（不阻塞但推荐）

无新增建议。Forrest-ly 此前两条建议（行 283 `??` 与行 304 `=== undefined` 的 content 判空语义不一致；文件类工具与 interrupted 结果必然触发一次 passthrough spawn）在当前 head 仍然存在且依旧成立，此处不重复。

### 🟢 值得肯定

- 三个已修复问题均落在正确位置且各有回归测试：`pwsh` 加入 `COMMAND_TOOLS`（行 38），非零退出按 `status: 'error'` + `command_output` + `replace_output: false` 发送；`commandFailed()` 将非空字符串 `signal` 判为失败（行 88），`failureText` 将 `terminated by signal: <signal>` 纳入诊断输入（行 127-129），SIGTERM + 空流场景不再产生空载荷；对象转文本边界恢复 `JSON.stringify` try/catch + `String()` 兜底（行 118-125），循环引用的下游 `value` 回归验证 listener 不 reject 且原值原样返回。
- `content_origin` 映射测试遍历共享 `tool_categories.json`，强制 layer_1/layer_2 工具映射为 `file_content`/`command_output`，同时允许 DSH 特有扩展（pwsh）——防漂移与宿主特异性兼顾得当。
- 失败开放路径覆盖全面（fail/invalid/timeout/missing/wrong-version/wrong-operation），`node --check dist/index.js` 本地通过；选项表、cordis.patch.yml 示例与代码中无已删除选项（skipTools、shellTools、truncateStringsAt、truncateArraysAt、maxDepth、noStash）的残留引用，旧 profile 携带这些键时仅被静默忽略，失败安全。
- 核对过 Core 侧路由（`entry.rs` post_tool_with_store）：error 状态的环境诊断与 `content_origin` 无关，因此文件类工具失败的环境归因在迁移后仍然成立；`retrieval_available: false` 使有损候选被拒，与文档"无 Marker 授权恢复路径"的描述一致。

**Review by Forrest-ly — APPROVED** [review]
## 总体评价

对新 head `29e7e911` 完成复审，重点核对 b2de411 → 29e7e911 增量（仅 `dist/index.js` 与 `tests/test-dsh-adapter.sh` 两个文件，+34/-33 行；文档/README/package.json.in 与 b2de411 逐字节一致）。此前唯一遗留的阻塞问题——"空 stderr 结构化命令失败绕过 Core"（ikunkun-sys 11:30 [P2]，非替换分支）及其在 replacesValue 分支的同一缺陷（Forrest-ly 11:48 review 🔴）——均已在本 head 修复并经我逐行与本地运行验证：两处 `if (!content) return decision` 早退均已移除，`failureText()` 新增 stdout 兜底且保持 stderr/error/signal 优先，所有宿主判定为失败的结果现在无条件送达 Core。此前三个已修复问题（pwsh 分类、signal 判失败、stringify 防护）在本增量中未被触碰，维持已验证状态。

## 审查结论

approve

## 详细意见

### 🔴 必须修改（阻塞合并）

无。核对依据：

- [`src/tokenless/adapters/tokenless/dsh/dist/index.js:267`、`314-316`] 两处失败分支均已删除空内容早退；`failureText()`（行 130-137）仅当已收集的所有部分均为空时才回退 `value.stdout`（string 或 `{ text }` 形态），stderr/error/signal 优先级不变——现有测试 "command stdout must not drive diagnosis"（test-dsh-adapter.sh 行 239-251）继续通过即证明了该优先级。
- Core 侧契约核对（`crates/tokenless-runtime/src/entry.rs` `post_tool_with_store`）：`status == Error` 路由为 passthrough + `Disposition::ToolError` + `diagnose_tool_error()`，并计入 EntryStats；空内容不会被协议（`deny_unknown_fields`）拒绝，`diagnose_tool_error` 对空串安全返回 None。因此"全空流失败"场景下 tool_error 路由与统计不再可被绕过，与文档声明一致。
- 回归测试到位：stdout-only（行 253-265，同时断言 `content_origin: command_output` 与 `replace_output: false`）与全空（行 267-278，断言请求仍以 `status: error`、`content: ''` 发出）两个用例恰好覆盖 ikunkun-sys 的复现场景。
- 本地验证：`node --check dist/index.js` 通过；`bash tests/test-dsh-adapter.sh` 全量通过。

### 🟡 建议修改（不阻塞但推荐）

- [`src/tokenless/tests/test-dsh-adapter.sh:322-355`] **（补充 Forrest-ly 11:48 review）** 该 review 建议的"替换值 + 空 stderr"回归未随修复提交：replacesValue 分支当前仅有 stderr 非空的两个用例（`connection refused` 替换值、循环引用 `error`），而新增的 stdout-only 与全空回归都落在非替换分支（行 253-278）。修复后两分支共享同一 `failureText()` 提取与"无条件发送"语义，共享逻辑已被钉住，但该分支直到本 commit 仍携带同一缺陷，建议补一条"下游替换值 + 空/全空 stderr"用例防止未来重构回归。不阻塞合并。
- 既有两条 🟡（行 290 `??` 与行 311 `=== undefined` 的 content 判空语义不一致；文件类工具与 interrupted 结果必然触发一次 passthrough spawn）及文档小建议（注明已删除选项不再被读取）在本 head 依旧存在且已被此前 review 覆盖，此处不重复。

### 🟢 值得肯定

- 修复取"无条件发送 + stdout 兜底"而非回退模型可见内容，是合理的最小设计：全空场景本无可诊断文本，而路由/统计保证不再依赖内容非空；stdout-only 场景（如 `permission denied` 走 stdout 后 exit 1）恰好能被 `diagnose_tool_error` 正确归类为 ENV_PERMISSION。
- stdout 兜底的作用域控制得当：仅在 `value` 对象分支内、且所有已收集部分为空时触发，不污染非结构化失败路径（`failureText(effectiveResult)` 仍取模型可见渲染内容）。
- 增量极小且聚焦（净 +1 行源码逻辑改动 + 两条回归），未触碰已验证的 pwsh/signal/stringify 修复，回滚与 bisect 友好。

**Review by ikunkun-sys — APPROVED** [review]
Re-reviewed current head `29e7e911`. The previous structured command failure gap is fixed: stdout-only and fully empty non-zero failures now always reach Core, including the downstream replacement path. Verified with the adapter tests, DSH plugin build, and real Core probes for ordinary and replacement-value failures. No blocking findings remain.

**Review by Forrest-ly — APPROVED** [review]
## Re-review of `2718aef` (post-APPROVAL force-push) — APPROVE

### 1. Squash purity — verified clean (the key re-review question)

Fetched both heads locally and diffed them directly:

- `git diff 29e7e91118089f94ea3e16b8a5d15523ab9c6210 2718aef4cbaccccb9800d31592786c1be20740de` → **empty** (byte-identical trees).
- Both commits share the same single parent `08d82c8b` (= merge-base with `main`); the branch is a single commit, author date (10:25:48Z) preserved, committer date restamped to 15:31:25Z.

The force-push is a pure squash of the doubly-approved head. **No semantic change.**

### 2. Full-diff review of `2718aef` (586+/684-, 9 files)

- `dist/index.js`: the bundle is now a thin lifecycle adapter. Verified: waterfall ordering preserved (`await next()` first, acts only on `accept` decisions); canonical values replaced by later listeners are never content-compressed and only receive diagnostics; `commandFailed` covers numeric/string exit codes, signal termination, timeout, and `isError/success/ok` flags; `failureText` uses stderr → error → signal note with stdout fallback and a circular-reference-safe `JSON.stringify`; DSH abort codes map to `interrupted` (distinct from tool errors); success path is gated by `responseCompressionEnabled`, skips Code Mode children, requires single-text content, and replaces content only on Core `disposition: 'applied'`; transport validation (`protocol_version === 2`, `operation === 'post_tool'`) with fail-open on every error path.
- Removed policy knobs (`skipTools`, `shellTools`, `truncateStringsAt`, `truncateArraysAt`, `maxDepth`, `noStash`) vs. retained (`responseCompressionEnabled`, `tokenlessBin`, `timeoutMs`, `maxBuffer`, `agentId`) match the bilingual docs exactly.
- `tests/test-dsh-adapter.sh`: comprehensive — request envelope shape, content-origin mapping cross-checked against `tool_categories.json`, block/mixed-content/canonical-value non-replacement, structured-failure diagnostics (stderr/stdout/empty precedence), `pwsh` classification, signal termination, interrupted vs error, Code Mode children, disabled-compression diagnostics, and the full fail-open matrix.
- Docs (en/zh framework-integration + user-manual, README/README_zh): consistent with the code; bilingual parity maintained. `package.json.in`: description-only change.

### 3. Local verification on this exact tree

- `node --check adapters/tokenless/dsh/dist/index.js` — OK.
- `bash tests/test-dsh-adapter.sh` — all assertions pass ("native DSH adapter tests passed").

### 4. Historical findings

All four prior findings (bot P2 pre-transform value snapshot; P1 `pwsh` classification; P2 signal-terminated commands; P2 structured command failures on the Core path) are verified fixed on this tree, including via the dedicated regression tests. The two threads not yet formally resolved on GitHub are housekeeping only — both fixes are present here.

No new findings. Approving.
